oeps/oep-0033-bp-code-autoformatting.rst

```diff
@@ -0,0 +1,127 @@
+===========================
+OEP-33: Code Autoformatting
+===========================
+
++-----------------+--------------------------------------------------------+
+| OEP             | :doc:`OEP-33 </oeps/oep-0033-bp-code-autoformatting>`  |
++-----------------+--------------------------------------------------------+
+| Title           | Code Autoformatting                                    |
++-----------------+--------------------------------------------------------+
+| Last Modified   | 2019-03-04                                             |
++-----------------+--------------------------------------------------------+
+| Authors         | Calen Pennington <cale@edx.org>                        |
++-----------------+--------------------------------------------------------+
+| Arbiter         | Ned Batchelder <ned@edx.org> (Tentatively)             |
++-----------------+--------------------------------------------------------+
+| Status          | Draft                                                  |
++-----------------+--------------------------------------------------------+
+| Type            | Best Practice                                          |
++-----------------+--------------------------------------------------------+
+| Created         | 2019-03-04                                             |
++-----------------+--------------------------------------------------------+
+| `Review Period` | 2019-03-04 - 2019-04-01                                |
++-----------------+--------------------------------------------------------+
+
+Context
+-------
+
+In a large software project or ecosystem like the Open edX codebase, consistency
+of code formatting makes it easier to read the code. In an ecosystem with many
+contributors, maintaining that consistency requires automated tools (such as
+linters), consistent code review, or both. In general, minimizing the number
+of review comments that a reviewer has to make on an incoming change makes
+review easier, and lessens the number of back-and-forths required to merge
+the code. As such, linters that can be run locally should be preferred.
+Similarly, ambiguity in style guidelines should be minimized, so that there
+are fewer places for commentary around what amounts to personal preference
+(which might then instigate unnecessary debate).
+
+In short, more consistent, less ambiguous standards for code formatting
+make for more productive code reviews and waste less time on inconsequential
+details.
```

Comment:
Have there been recent incidents where inconsistent formatting has resulted in longer PR reviews?  I have not found this to be a major issue in my own personal experience to-date working on our platform.

Reply:
Personally, I run into pylint violations that are simply a question of spacing/indentation fairly frequently. I also often refrain from commenting on formatting that would make the code more consistent because it's not something that I want to have another review cycle on. Maybe it's just me that's bothered by inconsistent code formatting?

Reply:
Ah, right.  I've also run into pylint violations, complaining of extra spaces that it doesn't allow. We could also just update our linter to ignore that violation if it's not necessary to enforce.

Reply:
My view is that there is value in consistency and in the spacing guidelines that pylint suggests, from a readability perspective, and that there isn't much value in making engineers actually enforce it themselves, since tooling can do that job faster and more effectively.

Reply:
Thanks @nasthagiri, great question. Yes, formatting and CI in general has been a painpoint for me while contributing to Open edX.

Despite its great value, it still puts a lot of unrelated issues to the code that I write -- say I edited line no. 1, it would complain about line no. 100. This is especially true for home-grown linters such as the XSS linter.

I'm not saying that we should loosen and "de-regulate" Open edX quality, but at the same time please don't underestimate the effort that infrequent contributors have to go through.

This is especially true for the `edx-platform` repository. 

I have learned my way through the process, but I find it hard to recruit other open source contributors whether from my company or from the community.

Reply:
I have a separate OEP (https://github.com/edx/open-edx-proposals/pull/108) that I want to use to cut off the accidental flagging from the XSS Linter. In general, I'm all for making it so that the only things that actually get flagged on a PR are substantive issues in code that you actually wrote in the PR.

Reply:
Thanks @cpennington! That would be great. Afterwards, we'd only be left with the inevitable cost of dealing with the CI on such a monolithic piece of code -- the `edx-platform`. As a long time (sine 2014) but infrequent contributor, I do see the value of the CI, but also don't want to undermine the cost of it to me as a contributor.

Therefore, I think we should look at such initiatives a bit more critically that we're doing now -- which is mostly focusing on the pro's of adding yet another step.

Reply:
Yes, the impact on time-to-feedback for edx-platform specifically is a good question.

On my machine (not devstack), w/ no black cache, it takes 10 minutes to format all of edx-platform. After reformatting everything, it takes 13s to reformat/check again.

After I reformat everything, and then delete the cache, a full format takes 5 minutes (and again, once cached, takes about 13s).

So, that suggests that a) we want to make sure jenkins keeps that cache around, and b) that devstacks are pre-built with the cache so that running black inside devstack to do a full-format (for someone w/o editor integration) is fast enough to not be a blocker.

Reply:
Sigh... turns out that black needs python 3.6 to run, which we don't have installed on devstack... So, we'd need some more investment to make edx-platform work.

Reply:
> I also often refrain from commenting on formatting that would make the code more consistent because it's not something that I want to have another review cycle on. Maybe it's just me that's bothered by inconsistent code formatting?

Internally at Stanford, I enforce this fairly rigidly, through sheer force of will :)
That doesn't always scale though...